doc/fundamentals/spouts.md

```diff
@@ -0,0 +1,287 @@
+
+In the example below, 
+written in Go
+and taken from the [kafka example](https://github.com/pachyderm/pachyderm/tree/master/examples/kafka)
+in the Pachyderm repo, 
+we'll go through every step you need to take.
+
+### Import necessary libraries
+
+We'll import the libraries necessary for creating a `tar` archive data stream
+and connecting to our Kafka data source.
+
+```go
+package main
+
+import (
+	"archive/tar"
+	"context"
+	"os"
+	"strings"
+	"time"
+
+	kafka "github.com/segmentio/kafka-go"
+)
+
+```
+### Parameterizing connection information
+It's a good idea to get parameters
+for connecting to your data source 
+from the environment
+or command-line parameters. 
+That way you can connect to new data sources
+using the same container code
+without recompiling
+by just setting appropriate parameters
+in the pipeline specification.
+```Go
+func main() {
+	// Get the connection info from the ENV vars
+	host := os.Getenv("HOST")
+	port := os.Getenv("PORT")
+	topic := os.Getenv("TOPIC")
+```
+
+### Connect to the streaming data source
+
+We're creating an object
+that can be used to read from our data source.
+
+That `defer` statement is the Go way to guarantee
+that the open file will be closed
+after the code in `main()` runs, 
+but before it returns.
+That is, 
+`reader.Close()` won't be executed
+until after `main()` is finished with everything else.
+(This is a common idiom in Go;
+to `defer` the close of a resource
+right after you open it.)
+```Go
+	// And create a new kafka reader
+	reader := kafka.NewReader(kafka.ReaderConfig{
+		Brokers:  []string{host + ":" + port},
+		Topic:    topic,
+		MinBytes: 10e1,
+		MaxBytes: 10e6,
+	})
+	defer reader.Close()
+```
+
+### Open /pfs/out for writing
+
+We're opening the named pipe`/pfs/out` for writing
+so we can send it a `tar` archive stream
+with the file
+we want to output.
+Note that the named pipe has to be opened
+with write-only permissions.
+```Go
+	// Open the /pfs/out pipe with write only permissons (the pachyderm spout will be reading at the other end of this)
+	// Note: it won't work if you try to open this with read, or read/write permissions
+	out, err := os.OpenFile("/pfs/out", os.O_WRONLY, 0644)
+	if err != nil {
+		panic(err)
+	}
+	defer out.Close()
+```
+
+### Write the outer file loop
+Here we open a tar stream into the directory we opened above,
+so that Pachyderm can place files into the output repo.
+For clarity's sake,
+we'll omit the message-processing loop inside this file loop.
+The `err` variable is used in the message-processing loop for errors reading from the stream and writing to the directory.
+The stream is opened at the top of the loop and should be closed at the bottom.
+In this case, 
+it'll be closed after a message is processed.
+The commit is finished after the file stream closes.
+(The `defer tw.Close()` is wrapped in an anonymous function to control when it gets run; 
+it'll run after all the code in the anonymous function is finished,
```

Comment:
i think it would be a little simpler and clearer to say "It'll run right before the anonymous function returns"